### DISCUSSION TITLE
I2C with Pi4j V2

### DISCUSSION THREAD

**Discussion by Brosbud** [Pi4J/pi4j#221] [Q&A]
Hi!

I'm having some trouble with getting the I2c to work.
I keep getting the following error messages;
  - PIGPIO ERROR: PI_BAD_I2C_BUS: bad I2C bus
  - PIGPIO ERROR: I2C/SPI/SERIAL HANDLE [-74]; Valid range >0

Is there something more that needs to be initialized? I'm following [this ](https://github.com/Pi4J/pi4j-examples/blob/master/src/main/java/com/pi4j/example/i2c/I2cDeviceExample.java) example

Heres some of my code;

`
The I2C bus is 1
The I2C device is 0x48 as per the PCF8591 module
The I2C A0 is 0x40

        var pi4j = Pi4J.newAutoContext();

        var config = I2C.newConfigBuilder(pi4j)
                .id("i2c-ADC")
                .name("PCF8591")
                .bus(I2C_BUS)
                .device(I2C_DEVICE) 
                .build();

        I2CProvider i2CProvider = pi4j.provider("pigpio-i2c");

        try ( var i2c = i2CProvider.create(config)) {

            
            var register = i2c.register(I2C_A0);
            register.write(0x0D);
            byte readByte = register.readByte();
            System.out.println(readByte);

        }
`

**Comment by jveverka** [comment]
you may check also:
https://github.com/jveverka/rpi-projects/tree/pi4j-v2/rpi-drivers
https://github.com/jveverka/rpi-projects/blob/pi4j-v2/rpi-drivers/src/main/java/one/microproject/rpi/hardware/gpio/sensors/tests/PCF8591Test.java

**Comment by taartspi** [comment]
I did a copy paste of the example you referenced.  That successfully  executes the steps shown above.  I changed the I2C_BUS value to 0 and see that same error you reported.            As I do not have a device ACKing that address I then did change the address value to match my attached devices as validated with  i2cdetect -y 1 and reset the I2C_BUS to 1.    It executes to completion.             Guess all I accomplished is validate the example works.   Maybe time to run with a debugger and determine if others changed you made created this condition.
